[WIP] Refactor/use oras and organize

vsoch opened the pull request:
This is a work in progress PR to refactor the channel mirrors to use oras. I will remove the WIP label when it is ready for review - I still need to refactor the cache_packages.py, add testing, and a few other things! I'm also opening on the fork so I can figure out testing before making too much noise on the main branch.

vsoch committed:
start of work to refactor conda_oci_mirror to use oras

I am adding a global oras client, that way we can ensure checking
credentials once (and possibly defaulting to a dry run) and using
the client across different types of pushes (repo and packages).
I also added a "Pusher" class to oras since the actions of creating
layers and setting defaults were used between the two places. I
am currently maintaining the relatively flat structure of the
module, however this could be refactored in the future if needed
(I usually prefer to have a main module with the core functionality
but I will not push this preference on others! I need to next
refactor the cache_packages.py and then add some testing.

Signed-off-by: vsoch <vsoch@users.noreply.github.com>

vsoch committed:
refactor of push/pull client commands

Signed-off-by: vsoch <vsoch@users.noreply.github.com>

vsoch committed:
testing adding CI to run basic commands

Signed-off-by: vsoch <vsoch@users.noreply.github.com>

vsoch committed:
add logger with levels and colors

Signed-off-by: vsoch <vsoch@users.noreply.github.com>

vsoch committed:
adding first test of validating mirror content

Signed-off-by: vsoch <vsoch@users.noreply.github.com>

vsoch committed:
add simple second push/pull test

this test needs to be extended but for now is setup to
do some push and some pull and validate.

Signed-off-by: vsoch <vsoch@users.noreply.github.com>

wolfv commented:
Finally had a chance to read through your changes and I am really excited. Very nice improvements! 

I do think two functions are missing from the previous "OCI" class -- to pull the index.json layer and to fetch the `info.tar.gz` as a `tarfile` (I might have overlooked something). It's not a big deal, but I think it's nice to have a higher level interface around a "conda package & tag" to fetch this specific metadata.

```py
    def get_info(self, package, tag):
        digest = self._find_digest(package, tag, C.info_archive_media_type)
        res = self.get_blob(package, digest, stream=False)
        return tarfile.open(fileobj=BytesIO(res.content), mode="r:gz")

    def get_index_json(self, package, tag):
        digest = self._find_digest(package, tag, C.info_index_media_type)
        return self.get_blob(package, digest).json()
```

In any case, I think this looks ready to be merged upstream!

wolfv commented:
Ah, just found the relevant code in `subdir_accessor.py` :) Forget my earlier comment, haha

wolfv commented:
I've tried your branch a bit more. There are two things I noticed:

- the repodata is pushed before we mirror the packages. That would mean that the repodata might contain some packages that are not (yet) available on the registry. I think it might be better if we do it at the end? Or at least tag with `:latest` at the end.
- The individual packages get a `latest` tag according to the logs I saw (github is weird and I don't see the actual packages yet, lol). E.g. these were my logs (testing with the .conda pacakges):

```
(base) ➜  cache git:(refactor/use-oras-and-organize) conda-oci mirror --subdir noarch -p intake-axds --user wolfv --debug 1
Using cache dir: /Users/wolfvollprecht/Programs/oci_mirror/cache/cache
Channels  : conda-forge
  Subdirs : noarch
  Packages: intake-axds
To:  ghcr.io/wolfv
  pushing tag 2023.01.28.11.10
⭐️ Pushing ghcr.io/wolfv/conda-forge/noarch/repodata.json:2023.01.28.11.10: 2023.01.28.11.10
Preparing layer {'mediaType': 'application/vnd.conda.repodata.v1+json', 'size': 78652347, 'digest': 'sha256:6b065ffb90045e9566c8e0e07c1b0c3943a99160bf18ed3cc5eaf26fc44292ee'}
Uploading repodata.json to ghcr.io/wolfv/conda-forge/noarch/repodata.json:2023.01.28.11.10
Successfully pushed ghcr.io/wolfv/conda-forge/noarch/repodata.json:2023.01.28.11.10
  pushing tag latest
⭐️ Pushing ghcr.io/wolfv/conda-forge/noarch/repodata.json:latest: 2023.01.28.11.10
Preparing layer {'mediaType': 'application/vnd.conda.repodata.v1+json', 'size': 78652347, 'digest': 'sha256:6b065ffb90045e9566c8e0e07c1b0c3943a99160bf18ed3cc5eaf26fc44292ee'}
Uploading repodata.json to ghcr.io/wolfv/conda-forge/noarch/repodata.json:latest
Successfully pushed ghcr.io/wolfv/conda-forge/noarch/repodata.json:latest
Rate limit sleep for 0.30303215980529785
Rate limit sleep for 0.2994699478149414
Rate limit sleep for 0.28709983825683594
Successfully pushed ghcr.io/wolfv/conda-forge/noarch/intake-axds:0.1.1-pyhd8ed1ab_0
Successfully pushed ghcr.io/wolfv/conda-forge/noarch/intake-axds:0.1.2-pyhd8ed1ab_0
Successfully pushed ghcr.io/wolfv/conda-forge/noarch/intake-axds:0.1.0-pyhd8ed1ab_0
Successfully pushed ghcr.io/wolfv/conda-forge/noarch/intake-axds:latest
Successfully pushed ghcr.io/wolfv/conda-forge/noarch/intake-axds:latest
Successfully pushed ghcr.io/wolfv/conda-forge/noarch/intake-axds:latest
```

I think for packages like these a "latest" tag doesn't really have meaning so would leave it out. There could be multiple variants of a given package with the same latest version number, for example, but only one of them could have a `latest` tag.

vsoch commented:
> I do think two functions are missing from the previous "OCI" class -- to pull the index.json layer and to fetch the info.tar.gz as a tarfile (I might have overlooked something). It's not a big deal, but I think it's nice to have a higher level interface around a "conda package & tag" to fetch this specific metadata.

The subdir_accessor.py is an artifact of the previous variant and I'm not sure how it was intended to be used. Now that I see it, I'll see if I can refactor it to be a part of the new design!

> the repodata is pushed before we mirror the packages. That would mean that the repodata might contain some packages that are not (yet) available on the registry. I think it might be better if we do it at the end? Or at least tag with :latest at the end.

I tried to honor the [previous design](https://github.com/wolfv/conda_oci_mirror/blob/155d67c2ad275876ef3fb523a273be1aa738df95/conda_oci_mirror/oci_mirror.py#L412-L417) that is currently being used, changing the structure without changing the function. We can definitely change the function to this! I agree it's probably better to be certain that the data is there before pushing it.

> The individual packages get a latest tag according to the logs I saw (github is weird and I don't see the actual packages yet, lol). E.g. these were my logs (testing with the .conda pacakges):

Ah, so in the tests we aren't actually pushing to GitHub packages - it's just a temporary registry that runs there so we wouldn't see them in packages! 

> I think for packages like these a "latest" tag doesn't really have meaning so would leave it out. There could be multiple variants of a given package with the same latest version number, for example, but only one of them could have a latest tag.

Could you elaborate on "packages like these?" And I am +1 for removing latest - it's a fairly contentious tag in the container world and given that there are means to already request a latest, we probably can leave it out.

Starting some work now!

wolfv commented:
Replying from mobile, can elaborate more later :) 

Regarding the latest tag, the package we're pushing has some versions and if I read correctly we're pushing three package versions and 3x the latest tag. However it's not clear which of the three packages actually receives the latest tag (of the three intake-axds versions from my logs).

For the repodata you're right, sorry. That bug was there before apparently :) for repodata the latest tag makes sense and we should keep it (used to fetch the latest repodata). We can change that later to push the repodata json file after mirroring of the packages is completed!

vsoch commented:
> Replying from mobile, can elaborate more later :)

No worries!

> Regarding the latest tag, the package we're pushing has some versions and if I read correctly we're pushing three package versions and 3x the latest tag. However it's not clear which of the three packages actually receives the latest tag (of the three intake-axds versions from my logs).

Haha yeah I just ran the tests again myself and noticed that, oups :P 

> For the repodata you're right, sorry. That bug was there before apparently :) for repodata the latest tag makes sense and we should keep it (used to fetch the latest repodata). We can change that later to push the repodata json file after mirroring of the packages is completed!

I'm trying out putting the repodata upload as a task, and it would be in the queue after the respective packages are done.

I'm in the middle of refactor and will ping after I've pushed some changes!

vsoch commented:
Quick update - I'm done with all of the above (will share details when I push) and I'm working on having the push/pull functions also be run with parallel tasks.

wolfv commented:
great! it's getting late here, but excited to take a look at your progress tomorrow.

vsoch committed:
add support for getting package, info index.json for a package

the previous subdir_accessor.py had functions to get info, the index.json,
and the package (archive) when given a specific URI. These are now added
to the PackageRepo, meaning this class can now take a registry URI (host
and username) on init so it functions similarity to the previous class.
To be more consistent across the cli and classes, we now take the full
registry <host>/<username> as the registry argument instead of separating
them. It also did not make sense for the Mirror class to accept multiple
channels when we are generally just taking one. Finally, the pull/push
classes now are done with parallel tasks, and the structure of tasks
is more consistent. There are tests for these new functions, and
the upload of a latest for a package is removed. This means when
we want to download latest, we determine the latest based on
sorted versions and build numbers

Signed-off-by: vsoch <vsoch@users.noreply.github.com>

vsoch commented:
>  great! it's getting late here, but excited to take a look at your progress tomorrow.

Haha indeed, time zone differences! It sounds like you are like me and stay up late :)

As promised, some updates! https://github.com/researchapps/conda_oci_mirror/commit/e86110046686e4d4c86ea8c1e6733b8377fae462

## Subdir Accessor

This class has been absorbed by `PackageRepo` that is largely the same abstraction.  They are now here: https://github.com/researchapps/conda_oci_mirror/blob/e86110046686e4d4c86ea8c1e6733b8377fae462/conda_oci_mirror/repo.py#L53-L102 and you can see how they are used in the tests (starting) here: https://github.com/researchapps/conda_oci_mirror/blob/e86110046686e4d4c86ea8c1e6733b8377fae462/conda_oci_mirror/tests/test_package_repo.py#L60. The main difference between the two was being able to provide the registry URI (e.g., `ghcr.io/wolfv`) upon init, and I've refactored the entire set of our classes to be consistent in this manner (next heading!) :point_down: 

## Registry argument

Comparing the previous class "SubdirAccessor" to the current classes, I realize that it's nice to be able to make a client (akin to the SubdirAccessor) that takes the registry from the getgo, and then you can just query for packages / tags. I originally had `registry` as a required param for upload / get tags or packages, and I've changed it to also accept the registry on init and then (optionally) a different registry for each of those functions, per discretion of the developer using it. This means that it can function very similarity to the previous `SubdirAccessor` (now removed).

This also made me go back and look at our main client functions, and they are expecting a user and host. We were being inconsistent with the usage of namespace/ registry / user. The original client accepted a user and registry, but the ultimate usage was to plop them together (e.g., `ghcr.io/wolff`). Since the other classes (e.g., PackageRepo that now has functions for SubdirAccessor) just accept the full registry, let's be consistent and keep it simple and accept that one argument. It will change the client interaction than what is currently there, but better earlier than later I think. E.g,.:

```diff
- $ conda-oci push-cache --host ghcr.io --user researchapps --dry-run --package zlib --subdir linux-64
+ $ conda-oci push-cache --registry ghcr.io/researchapps --dry-run --package zlib --subdir linux-64
```

## Tests

New tests are added for the new functions for the PackageRepo, and I've removed redundant code with some shared functions in `helpers.py`

## Mirror channel

In that same thread to the above, I think a Mirror should also point to just one channel. The client _could_ still accept more than one, but it will create individual Mirror instances to interact with (and then we can also separate the upload logic so, for example, if one Mirror channel fails, we can choose to still try the others or abort at some point). The current design is actually [already doing this](https://github.com/wolfv/conda_oci_mirror/blob/a1b8f916bdfa7f78a90cf05281aab7651526f705/conda_oci_mirror/cli.py#L44) so I think we should follow that. TLDR It's a more modular design.

## Latest 

I removed the latest tag for a package upload - it didn't make sense This means, however, when we ask to download latest, we determine that by first getting the highest `build_number` for a given package, and then sorting the versions to get the latest.

## Tasks

I cleaned up the design a bit so they are more obviously named, e.g., `RepoUploadTask` vs `PackageUploadTask` and a few other tweaks like accepting the package verbatim instead of instantiating it.

There are likely more tweaks I could do (I was honoring the original design as much as possible) so if something sticks out as wrong, please bring it up and we can tweak it further! 

It looks like you've given me some guidance [here](https://github.com/researchapps/conda_oci_mirror/pull/3/files) about the new package format. Should this be the new default and the old one discarded?

wolfv commented:
Yeah, I'll add some comments regarding repodata and the new package format! 

- The new package format seems to work pretty much out of the box (since we already use the conda-package-handling api!). So no big changes needed to get `info.tar.gz` and `index.json` support
- We just need to "find" the new packages and that's done by looking at the `packages.conda` section of the repodata.json. We should continue to mirror the old version and the new version as (at least in the case of conda-forge) they do not overlap (e.g. new versions of packages only appear under `packages.conda`). That's what I did in that PR and tested and seemed to work pretty good :)
- Another detail about repodata.json: the repodata can and will be "patched" by patches derived from the https://github.com/conda-forge/conda-forge-repodata-patches-feedstock – so another improvement would be to use the `repodata_from_packages.json` file instead of the `repodata.json` file for the actual mirroring to also find & mirror "yanked" versions (but still also push the patched `repodata.json` version, because that one will be used for the actual package resolution by mamba!).

wolfv commented:
I pushed the `intake-axds` package here (on my personal account) to test. I used that package because it has few versions and all of them are in the new `.conda` format. Seems to work fine! https://github.com/users/wolfv/packages/container/package/conda-forge%2Fnoarch%2Fintake-axds

vsoch commented:
Quick question - how do I interact with your remote, e.g., https://github.com/users/wolfv/packages/container/package/conda-forge%2Fnoarch%2Fintake-axds if I can't push and there is no repodata.json? The first (not being able to push) seems OK because I could do `pull-cache` to get your packages locally, but to do that it's going to first look for the repodata.json (and there is none). Is it the case that it should be there, or I should be using repodata.json from elsewhere? :thinking:

vsoch commented:
For the time being I'm going to look for a packages.conda that's on the main channel-mirrors - I think maybe your testing didn't push the repodata.json (and maybe this is a bug?)

Update: found "zope.event" has both, at least for noarch!

```
zope.event-4.4-pyh9f0ad1d_0.tar.bz2
zope.event-4.5.0-pyh9f0ad1d_0.tar.bz2
zope.event-4.6-pyhd8ed1ab_0.conda
```

vsoch commented:
@wolfv before we merge this, I think it would be helpful to have a small set of paragraphs that describe the explicit actions for each of mirror / pull-cache, and push-cache so I can sanity check what I'm doing vs. what is expected. I've made a lot of changes and I'm worried about, for example, the cases where we are supposed to be using local vs. remote repodata.json. E.g., as an example, "push-cache" is checking for new packages as determined by the local repodata.json, and not a remote repodata.json. I'm not sure if that is right - I think we would probably want to check the local packages against the remote repodata, but maybe I'm wrong about that.

vsoch committed:
adding support for new .conda package format

Signed-off-by: vsoch <vsoch@users.noreply.github.com>

wolfv commented:
OK, so my mirror is indeed not very "functional" but still serves as a backup of some packages. Let me explain a little bit on how this is intended to be used and operated for the mamba / conda-forge use case:

Initially, conda-forge will continue to push packages to anaconda.org and we will periodically fetch all new packages from anaconda.org and mirror them to the OCI registry (e.g. every 6 hours or so). In this use-case we would consider the anaconda.org channels as authoritative and directly use their repodata.json (and not regenerate any repodata).

mamba will be able to look for the `repodata.json:latest` file and download that to get the repodata index. The repodata.json file contains the `sha256` sum of each artifact. We can use the sha256 sum and the name of a given package to pull from the OCI registry (we can pull the blob directly without going through the manifest, since we already know the sha256 from the repodata).
Furthermore, mamba would also be able to get the `repodata.json` from anaconda.org but the packages from the OCI registry (since the OCI registry and anaconda.org will be mirrors).

My package-only mirror doesn't really work straight away for mamba (except if we have repodata.json from anaconda.org), but it could serve as a personal backup or with a lockfile (where we store the sha256 and URL of the package concretely and don't need a repodata.json to resolve).

It could also just be used to retrieve some metadata without downloading the whole file, like the functions we have now to get info or index.json. But it was mainly thought for testing.

Later, it could be very interesting to have per-package repodata.json files btw.

vsoch commented:
@wolfv I added support for .conda! https://github.com/researchapps/conda_oci_mirror/pull/2/commits/10899e0570244bf442656dfdab0ebe2f1b4a7271. Before I add `repodata_from_packages.json` I want to suggest we do a small summary of what each command is supposed to do - I just added some small documentation to the README about the endpoints, and in particular I want to clarify the pull-cache and push-cache. Intuitively, I'd think these functions are going to update my cache with my remote registry, but the push-cache seems to only compare the local repodata with what is in the cache (and not in the remote) so I think I want to clarify that, and then ask (depending on which is done) how we would support the other interaction. E.g., maybe we want something like a sync-cache? It seems like there are a few options (and here is my current interpretation of what the commands should do):

- a mirror is saying "take the channel on the official conda and push it to my registry"
- a pull-cache is saying "get the latest packages from my registry that I don't have locally, and get them"
- a push-cache is updating my local repodata with the files I actually have, and uploadling ones that are deemed new.

But for the last case, what about packages in my local cache that aren't deemed new, but I don't have in my registry? That seems like a slightly different case, and maybe we should instead have:

- a sync-cache is updating my local repodata with the files I actually have, comparing that to the remote, and uploading anything that doesn't exist there.
- a push-cache is doing the same, but only uploading the ones that are deemed new locally.

Or something like that?

Also tests were hanging (and passing locally) so I might need to push again with a bugfix (if there is one) - I'll update the link at the top if I do.

Update: they passed this time, GitHub must have been being derpy. Like me earlier when I maximized my terminal in VSCode and had no way to minimize it again. :laughing: :sob:  It's funny now :)

vsoch commented:
> Later, it could be very interesting to have per-package repodata.json files btw.

I really like this idea - it would be very nice to be able to look for a repodata.json specific to one package, and have a more modular approach for interacting with a remote and the cache. What we couldn't easily do is to get a listing of all packages in a repository, but maybe the solution would be to have both (if that doesn't risk them getting out of sync!)

wolfv commented:
Yeah, that sounds like a good idea :) Actually, I had to remind myself a little bit what push and pull cache were supposed to do, haha.

So the `mirror` command should truthfully mirror all packages from a regular conda-channel to an OCI registry. This means, copying verbatim the existing repodata.json and all (selected) packages. The only "proper" way to use this is to copy all packages, I am not sure if a subset of packages will really be interesting except for testing.

The push- and pull-cache use case (so far) was to act as a cache for building out whole package trees. Admittedly there have been only two use cases so far: [boa-forge](https://github.com/orgs/mamba-org/packages?repo_name=boa-forge) and [emscripten-forge](https://github.com/emscripten-forge/recipes). In both cases we have a bunch of recipes and build some packages. Now, with the push-cache we can push new (intermediary) packages to the cache and re-use quite easily. The OCI registry is not strictly necessary (we could also push to github cache, for example). But a nice property of the OCI registry is that we can also use/debug the packages outside of Github, e.g. locally.

The way this was implemented in our prototypes was that we push the latest changes based on a comparison of what we have locally (after building all recipes) and what we had before. In the prototype we were indeed also attaching a `latest` tag to the packages (because were not building any variants) and for restoration we pulled the latest tags.
However, I think an approach with repodata.json will be much better. Pulling the cache explicitly might also not be necessary anymore in the future if we can point mamba to the build cache, and it will just fetch what it needs (as if it was a regular conda channel).

Sorry for not specifying this use case better beforehand. The mirroring is definitely higher priority for now :)

vsoch commented:
Thanks for the clarification! Here are some points for discussion.

## Repodata Sync

When we run a mirror, if the repodata.json already exists, we also don't try to update it, e.g., if we have made changes locally. This also won't get remote changes. So I think what we likely should do is always retrieve it to update (no matter what). If there are changes locally (possible) we could either 1. issue a warning and encourage the user to run a command to update and push, 2. take the position that the local cache should always be in sync with the remote and do nothing, or 3. try to do some kind of sync. I think actually the second is my preference - a mirror should be just that, a mirror of the upstream.

## Testing

Is there any kind of channel / package(s) that could explicitly be used for testing? We've chosen packages that meet certain criteria (e.g., zlib has lots of versions, I found one with .conda) but they are a bit brittle for testing in that they change and I can't select specific metadata to sanity check (it's a moving target). It's probably OK for now, but maybe for the future we could think about a testing setup for this?

wolfv commented:
I think when we're mirroring we should always just push the repodata that we've retrieved, yeah.
It should be an exact copy.

Regarding testing, we can setup a small fixed channels with a bunch of packages! I'll give it a go tomorrow :)

vsoch committed:
add better clarification to README, and always retrieve new repodata.json

Signed-off-by: vsoch <vsoch@users.noreply.github.com>

vsoch commented:
okay I did a little more tweaking, I added support for a serial mode (for debugging) and also download the repodata.json on mirror every time, no matter what. I think we can further improve upon the push/pull functionality, but it's not high priority compared to mirror so it can wait. For testing, I think as soon as we can get that reliable set of testing packages to mirror (and then I update the tests here) we should be pretty good (and I can look at that other repodata file).

Update: the testing set is essential, we just had another flaky "freeze and I'm not sure why" run. I'll make sure to have the actions open so I can see what's going on next time / if it happens again.

wolfv commented:
I made a small testing channel with a collection of packages in both `.conda` and `.tar.bz2` formats. There are also some `noarch`, `linux-64`, `osx-arm64` and `osx-64` packages: https://anaconda.org/mirror-testing

In total it should be less than 5 Mb though, 20 packages or something.

vsoch committed:
update tests to use testing mirror

Signed-off-by: vsoch <vsoch@users.noreply.github.com>

vsoch commented:
@wolfv I have loaded each of repodata.json and repodata_from_packages.json. The first has a 'removed' section with a list of archives, and the second (patches?) does not. E.g.,

```python
len(patches['packages'])
Out[9]: 115909

len(repodata['packages'])
Out[10]: 115639
```

You mentioned:

> so another improvement would be to use the repodata_from_packages.json file instead of the repodata.json file for the actual mirroring to also find & mirror "yanked" versions (but still also push the patched repodata.json 

So if I understand correctly, we want to use the larger list (without those removed) for the mirror, but when we push the actual repodata.json, we want to keep it as is. Is that correct? And the goal would be to ensure that yanked packages are available in the mirror too? But just to double check - there isn't any kind of "patching" I actually need to do between these two files.

vsoch committed:
use repodata_from_packages.json if exists

This file includes yanked packages, so we use it to do the mirror,
although we upload and honor the current repodata.json. In addition,
I am adding a helper script to run the registry so the developer
does not need to remember / copy paste from the readme, and support
for running in serial to the other mirror functions to be consistent!

Signed-off-by: vsoch <vsoch@users.noreply.github.com>

vsoch commented:
okay check out what I just added - I'm not sure I like my variable naming, but the (for now) `include_yanked` will use the repodata_from_packages.json instead for the mirror, while still ultimately uploading the repodata.json. https://github.com/researchapps/conda_oci_mirror/pull/2/commits/42696079a1d7a20ca591515546a74d69c51972a5. It's not exposed to the cli yet, but I suspect would /might be used within Python somewhere. Better naming suggestions appreciated!

wolfv commented:
Yes, the patching (currently) happens on anaconda.org servers and I think we should just mirror them for now.

The patching will change some dependencies of packages and also move some packages in repodata to the `removed` section. In repodata_from_packages the dependencies and packages are "as they are" in the packages themselves... so an easier source of truth for mirroring the packages, but `repodata.json` is what should be used at the end of the day for resolving stuff :)

It's nice to also have yanked packages available because people that have lockfiles might pull in some broken packages as well... and then the lockfiles continue working even though the packages aren't used for new environments anymore.

vsoch commented:
Gotcha! So next I was thinking of tackling shallow vs. deep modes [reference](https://github.com/researchapps/conda_oci_mirror/pull/2/files#diff-5c6a1301c6b59b30a040d747d065e861d3dd98bde0e5a4356d92d594e9835986R9-R11). You mentioned the [latest repodata](https://github.com/orgs/channel-mirrors/packages/container/conda-forge%2Fnoarch%2Frepodata.json/65546769?tag=latest) - are you saying to skip checking the repodata from conda.anaconda.org and just get it from a registry? Why is that more shallow?  I think I'm probably just missing a detail here :laughing: 

And then the other TODO was the compressed repodata.json as repodata.json.zst. If it's not in mamba yet, should we generate it and mirror both repodata.json and the .zst and eventually start using it? I suppose it wouldn't be used anywhere in a mirror, at least yet, but possibly when we do a pull we could use it instead for a slightly faster interaction.

And finally, when we like all the features it would be good to test the multiprocess upload on an actual registry. I'm not confident in the current backoff / ratelimit times, or at least maybe once in a development cycle I see a "405 method not allowed" when it's running with the workers and I want to dig into that (if it's my dummy registry or an issue with the implementation).

wolfv commented:
So in the initial prototype we determined which packages to mirror based on listing the tags of the uploaded packages. That takes quite a while to check 10k packages for all their tags (you can see a github actions run here: https://github.com/channel-mirrors/mirrormirror/actions/runs/3999269628/jobs/6862956904 (14 minutes for mostly checking tags, and that's only for one subdir).

So instead of checking tags, we could use the `repodata.json` (or `repodata_from_packages.json`) that we uploaded the last time around (assuming everything was fine and all tags were indeed created!).

So for the mirroring we'd download the `repodata.json:latest` from our registry and the `repodata.json` from anaconda.org and compare those two files, uploading all files that are missing from the `repodata.json` on the registry. 

The "deep" mode would be the one that still checks all tags to make sure that files are actually there.

The `repodata.json.zst` files are there for certain channels (e.g. https://conda.anaconda.org/conda-forge/linux-64/repodata.json.zst). We could verbatim copy it and upload it as another layer for the `repodata.json` push with a `vnd.conda.repodata.v1+zstd` or so ... actually I think we haven't come up with a filetype for repodata yet? ...

wolfv commented:
But also happy if we do that in a separate PR and run your changes in the current GitHub action after merging first?

vsoch commented:
Oh got it! Yep I can do both of those things.

So merging is totally up to you - I can keep working on this branch or start a new one tomorrow when I get some time after work. I’m turning into a pumpkin in T minus 10 minutes but I’ll start off with whatever you decide to do tomorrow evening. Thanks for the details to the above and have a good work day! 🧳

wolfv commented:
I'll wait for you to open the PR to the real repo. Also doesn't matter so much to me but always happy to press green merge buttons :)

vsoch commented:
okay see https://github.com/wolfv/conda_oci_mirror/pull/1. I think this will merge into the branch that is currently being used, let me know if we should do something different!

vsoch commented:
okay also opened https://github.com/channel-mirrors/conda_oci_mirror/pull/19 - I think I must have tried to base it on that main branch (and not wolfv) because there was no rebasing necessary there - but honestly I can't remember at this point! :laughing:

vsoch committed:
update TODO we already added support for .conda

Signed-off-by: vsoch <vsoch@users.noreply.github.com>
